### ISSUE TITLE
Doc issues: 

### ISSUE THREAD

**Issue by mhevery** [angular/angular#18141]
- [ ] I have to type 3 characters in https://angular.io/api?query=on before the list is updated. This makes it hard to search for all `on` interfaces.
- [ ] https://angular.io/guide/lifecycle-hooks should have a link to https://plnkr.co/edit/mtSIfJzHkBn2XuKqtfxZ?p=preview so that people can experiment.

**Comment by sergeome** [comment]
Hey, @mhevery, @Foxandxss.

I wanted to fix this bug. However, I have some questions:
1. API search query. Currently, it works for two characters, but it also shows the results if section title matches the search query. For example:

```
mySectionTitle

test
t-test
pot-test
```

None of these items contain `on`, but `mySecti'on'Title contains`, so a whole section will be displayed.

Would it be fine if I would display a whole section only if this section contains items that match the search query?

In addition to that, right now results are shown if a section item contains the query. It can be in the beginning, middle or end of the word. I can make search results visible only if it matches the beginning of the section item. For example:

```
foo
bar-foo
```

By typing foo it will display only `foo`. Do you need that?

2. Lifecycle-hooks section already contains a link to [plnkr](https://angular.io/generated/live-examples/lifecycle-hooks/eplnkr.html). Do you still want to include the link you mentioned in this issue? If so, could you please mention a place in the doc.

Please correct me if I'm wrong with something here.
Best regards,
Sergey

**Comment by gkalpak** [comment]
I am not a big fan of matching at the beginning of the word (at least not as the only option).

Alternatives/ideas that I prefer:

1. Make it configurable (e.g.  add a checkbox for matching anywhere in the word vs at the beginning only).
2. Allow using RegExp :smiley:

Alternatives/ideas that I also don't like (but did consider briefly :grin:):

1. Allow using `*` as a wildcard. (The feature needs to be brought to the user's attention and it makes it more verbose to match anywhere in the word.)
2. Matching at the beginning of each "sub-term". E.g. when matching the search query against `AsyncPipe` we would try to match at the beginning of `Async`, but also at the beginning of `Pipe`. (Too complex and too much magic under the hood.)

**Comment by petebacondarwin** [comment]
I don't think we should only match at the beginning of the word. In general people will put in larger query terms which will only capture real results and often we want things that are not just at the start. Imagine searching for `client`, you would probably want to get `HttpClient` no?

**Comment by petebacondarwin** [comment]
I am also keen to steer away from magic queries. At least showing too many queries in this case is consistent...

**Comment by petebacondarwin** [comment]
We are going to split the API elements by capital letter into parts and then only match the query to parts that start with the query.

**Comment by sergeome** [comment]
@petebacondarwin,

Thank you for the update!

**Comment by petebacondarwin** [comment]
@sergeome - did you want to have a go at doing this?

**Comment by sergeome** [comment]
@petebacondarwin, as far as I understand because you assigned this issue on yourself you already start working on it, right? In addition to that, I'm not sure If this "split idea" completely makes sense for me and I would have a good algorithm for this.
